DxDispatch/src/model/ImageReaderWriter.cpp

```diff
@@ -0,0 +1,271 @@
+
+    for (size_t pixelIndex = 0; pixelIndex < tensorInfo.height * tensorInfo.width; pixelIndex++)
+    {
+        for (size_t channelIndex = 0; channelIndex < tensorInfo.channels; channelIndex++)
+        {
+            // dst is always HWC layout.
+            size_t dstOffset = pixelIndex * tensorInfo.channels + channelIndex;
+
+            // src is either HWC or CHW layout.
+            size_t srcOffset = tensorInfo.layout == ImageTensorLayout::NHWC ?
+                dstOffset :
+                pixelIndex + channelIndex * tensorInfo.height * tensorInfo.width;
+
+            dst[dstOffset] = static_cast<std::byte>(std::max(0.0f, std::min(1.0f, (float)src[srcOffset])) * 255.0f);
+        }
+    }
+
+    return dst;
+}
+
+void WriteTensorToImage(std::filesystem::path dstPath, std::span<const std::byte> srcData, const ImageTensorInfo& srcTensorInfo)
+{
+    WICPixelFormatGUID desiredImagePixelFormat = GUID_WICPixelFormatDontCare;
+
+    switch (srcTensorInfo.channelOrder)
+    {
+        case ImageTensorChannelOrder::RGB: desiredImagePixelFormat = GUID_WICPixelFormat24bppRGB; break;
+        case ImageTensorChannelOrder::RGBA: desiredImagePixelFormat = GUID_WICPixelFormat32bppRGBA; break;
+        case ImageTensorChannelOrder::BGR: desiredImagePixelFormat = GUID_WICPixelFormat24bppBGR; break;
+        case ImageTensorChannelOrder::Grayscale: desiredImagePixelFormat = GUID_WICPixelFormat8bppGray; break;
+        default: throw std::invalid_argument("Unsupported channel order");
+    }
+
+    std::vector<std::byte> pixelBuffer;
+    switch (srcTensorInfo.dataType)
+    {
+        case DML_TENSOR_DATA_TYPE_FLOAT32: pixelBuffer = ConvertTensorToPixels<float>(srcData, srcTensorInfo); break;
+        case DML_TENSOR_DATA_TYPE_FLOAT16: pixelBuffer = ConvertTensorToPixels<half_float::half>(srcData, srcTensorInfo); break;
+        default: throw std::invalid_argument("Unsupported data type");
+    }
+
+    std::span<BYTE> byteSpan(reinterpret_cast<BYTE*>(pixelBuffer.data()), pixelBuffer.size());
+
+    ComPtr<IWICImagingFactory> wicFactory;
+    THROW_IF_FAILED(CoCreateInstance(
+        CLSID_WICImagingFactory,
+        nullptr,
+        CLSCTX_INPROC_SERVER,
+        IID_PPV_ARGS(&wicFactory)
+    ));
+
+    ComPtr<IWICBitmap> bitmap;
+    THROW_IF_FAILED(wicFactory->CreateBitmapFromMemory(
+        srcTensorInfo.width,
+        srcTensorInfo.height,
+        desiredImagePixelFormat,
+        srcTensorInfo.width * srcTensorInfo.channels,
+        static_cast<uint32_t>(byteSpan.size()),
+        byteSpan.data(),
+        &bitmap
+    ));
+
+    ComPtr<IWICStream> stream;
+    THROW_IF_FAILED(wicFactory->CreateStream(&stream));
+    THROW_IF_FAILED(stream->InitializeFromFilename(dstPath.wstring().data(), GENERIC_WRITE));
+
+    GUID containerFormat;
+    if (dstPath.extension() == L".jpg")
+    {
+        containerFormat = GUID_ContainerFormatJpeg;
+    }
+    else if (dstPath.extension() == L".png")
```

Comment:
Will it handle `FOO.PNG`? Is case already lowercased by the caller? I see you handle it properly in `GenerateInitialValuesFromFile` with:

```
    std::transform(fileExtension.begin(), fileExtension.end(), fileExtension.begin(), 
        [](unsigned char c) { return std::tolower(c); }
    );
```